lib/puppetserver/ca/action/generate.rb

```diff
@@ -199,21 +199,35 @@ def generate_certs(certnames, alt_names, settings, digest)
 
             current_alt_names = process_alt_names(alt_names, certname)
 
-            key, csr = generate_key_csr(certname, settings, digest, current_alt_names)
-            next false unless csr
-            next false unless ca.submit_certificate_request(certname, csr)
-            next false unless ca.sign_certs([certname])
-            if result = ca.get_certificate(certname)
-              next false unless save_file(result.body, certname, settings[:certdir], "Certificate")
-              next false unless save_keys(certname, settings, key)
+            next false unless submit_csr(certname, ca, settings, digest, current_alt_names)
+
+            # Check if the CA autosigned the cert
+            if download_cert(ca, certname, settings)
+              @logger.inform "Certificate for #{certname} was autosigned."
               true
             else
-              false
+              next false unless ca.sign_certs([certname])
+              download_cert(ca, certname, settings)
             end
           end
           passed.all?
         end
 
+        def submit_csr(certname, ca, settings, digest, alt_names)
+          key, csr = generate_key_csr(certname, settings, digest, alt_names)
+          return false unless csr
+          return false unless ca.submit_certificate_request(certname, csr)
+          return false unless save_keys(certname, settings, key)
```

Comment:
Ahhh, I tried setting autosign to true and allow-auth-extensions to false in my puppetserver.conf and then doing a generate with auth extensions and this still fails but it looks like it fails because an assumption of mine was wrong:
```
puppetserver-ca-cli justin (mergemaggie $) ‖ be exe/puppetserver-ca generate --certname foo
Error:
    When attempting to submit certificate request for 'foo', received:
      code: 400
      body: CSR 'foo' contains an authorization extension, which is disallowed. To allow authorization extensions, set allow-authorization-extensions to true in your
puppetserver.conf file, restart the puppetserver, and try signing this certificate again.
```
That's the csr submission failing rather than the sign failing.
Looking at the CA
```
puppetserver-ca-cli justin (mergemaggie $) ‖ tree ~/.puppetlabs/etc/puppet/ssl/
/Users/justin/.puppetlabs/etc/puppet/ssl/
├── ca
│   ├── ca_crl.pem
│   ├── ca_crt.pem
│   ├── ca_key.pem
│   ├── ca_pub.pem
│   ├── infra_crl.pem
│   ├── infra_inventory.txt
│   ├── infra_serials
│   ├── inventory.txt
│   ├── requests
│   │   └── foo.pem
│   ├── root_key.pem
│   ├── serial
│   └── signed
│       └── localhost.pem
├── certificate_requests
├── certs
│   ├── ca.pem
│   └── localhost.pem
├── crl.pem
├── private
├── private_keys
│   └── localhost.pem
└── public_keys
    └── localhost.pem
```

Unfortunately, the master has still saved the csr. I'm not sure if we should save the keys even if the submission fails, or if we should try to figure out if the request failed in a way that would leave the submission on the CA? Or if we should go back to requiring the cleaning of CSRs after any error in the process.

Reply:
I see two options here:
1) Does it make sense for puppetserver to save the CSR in this case? Even though it rejected it as a "bad request"? Maybe it shouldn't save it.
2) If we know what codes result in the CSR being saved, treat all of those as a "success" here. We wanted to submit a CSR, and we did.
I'm leaning towards (2), or else (1) while also updating puppetserver's error messaging to point out that the CSR will need to be resubmitted. I don't really want to go back to requiring cleaning of CSRs, because I think it would unclear to me as a user whether I'd also need to clean keys or do other things. If a CSR exists, it seems like a reasonable assumption that the keys would exist too...

Reply:
It looks like we return either a 2xx, 400, or 500 (from [the handler](https://github.com/puppetlabs/puppetserver/blob/master/src/clj/puppetlabs/services/ca/certificate_authority_core.clj#L57)). Of the reasons we return 400 we'll have duplicate certs and invalid certnames will cause the cert to not be persisted (makes sense since we save it to a file based on the certname), other 400 codes represent problems with autosigning and will leave the csr on disk.

Reply:
I've been thinking that agent-server comms should return 200 if it's a new CSR or the same CSR as what the agent already sent. Otherwise return 422 if the agent sends a different CSR with the same name, or 400 if the agent sends a bad request/illegal extension. The 422 means the server understands the request and it's valid, but the server can't process the request and the client shouldn't attempt to resend it.

Reply:
But that doesn't really help here, right? We still have some 400s that save the CSR, and others that don't. So I'm not sure we can be smart about this...?

Reply:
Ah okay. So would 2xx extend to cases with altnames/extensions, since ultimately the CSR submission is fine if the server chooses to ultimately change its settings and sign it?

Reply:
That makes sense to me, only return success so the agent knows it can transition from the "I have a local CSR" state to the "CSR accepted, awaiting cert" state.

Reply:
I don't think we can change this behavior now though. These are more good things to keep in mind for a v2 of the CA API.

Reply:
So in the meantime I think I'm just going to change it to save the keys in all cases. If the CSR failed to submit, hopefully the server will present a useful error message saying so. If the user then tries to run `generate` again, we can tell them to use `puppet ssl clean` to clean up the old keys.